# Table of contents

1. Importing Libraries
2. Import dataframe
3. Grouping data
4. Aggregating data with transform
5. Calculating descriptive statistics
6. Calculating average price
7. Calculating median price
8. Export to pickle - orders_products_merged_new



# 1. import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# 2. import data

In [2]:
# allocate file path

path = r'C:\Users\admin\Desktop\10.2023 Instacart Basket Analysis'

path

'C:\\Users\\admin\\Desktop\\10.2023 Instacart Basket Analysis'

In [5]:
df_ords_prods_merged=pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [6]:
ords_prods_merge =df_ords_prods_merged

# Task 2

In [7]:
ords_prods_merge.shape

(32404859, 19)

In [8]:
# step 2: Group entire dataframe by 'department_id' and calculate the mean of 'order_number' for each group.

ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

# task 3

The table shows the list of all possible 'department_id' values from the entire dataframe with their corresponding 'order_number' means - in other words, the average number of orders per user for each department ID. the department with id of 21 has a mean of around 23, which is the highest.

# Task 4

# Aggregating Data with transform()

In [9]:
# Create a new column 'max_order' in the 'ords_prods_merge' DataFrame that contains the maximum 'order_number' for each 'user_id'.

ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [10]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest day     busiest_day busiest_slowest_days busiest_period_of_day  \
0  Regularly busy  Regularly busy       Regularly busy        Average orders   
1  Regularly busy  Regularly busy         Slowest days        Average orders   
2  Regularly busy  Regularly busy         Slowest days           Most orders   
3      Least busy      Least busy         Slowest days        Average orders   
4      Least busy      Least busy         Slowest days           Most orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10

In [ ]:
# Configure pandas to display an unlimited number of rows, removing the limit on row display for DataFrame outputs.

# this code will display 100 rows


pd.options.display.max_rows = None

In [11]:
# Categorize customers based on their 'max_order' value into 'Loyal customer', 'Regular customer', or 'New customer' 
# and assign these labels to a new column 'loyalty_flag'.

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

# step 5

In [12]:
# Group the dataframe by 'loyalty_flag' and calculate descriptive statistics for 'price'.
loyalty_price_stats = ords_prods_merge.groupby('loyalty_flag')['prices'].describe()

# Display the statistics.
print(loyalty_price_stats)

                       count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal customer    10284093.0  10.386336  328.017787  1.0  4.2  7.4  11.2   
New customer       6243990.0  13.294670  597.560299  1.0  4.2  7.4  11.3   
Regular customer  15876776.0  12.495717  539.720919  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal customer    99999.0  
New customer      99999.0  
Regular customer  99999.0  


    Loyal customers have the lowest average spend at $10.39.
    New customers have the highest average spend at $13.29.
    Regular customers are in the middle at $12.50.

In [13]:
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [14]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  department_id  prices  _merge    price_range_loc  \
0         Soda  ...              7     9.0    both  Mid-range product   
1         Soda  ...              7     9.0    both  Mid-range product   
2         Soda  ...              7     9.0    both  Mid-range product   
3         Soda  ...              7     9.0    both  Mid-range product   
4         Soda  ...              7     9.0    both  Mid-range product   

      busiest day     busiest_day busiest_slowest_days busiest_period_of_day  \
0  Regularly busy  Regularly busy       Regularly busy        Average orders   
1  Regularly busy  Regularly busy         Slowest days        Average orders   
2  Regularly busy  Regularly busy         Slowest days           Most orders   
3      Least busy      Least busy         Slowest days        Average orders   
4      Least busy      Least busy         Slowest days           Most orders   

  max_order  loyalty_flag  
0        10  New customer  
1        10  New customer  
2        10  New customer  
3        10  New customer  
4        10  New customer  

[5 rows x 21 columns]

# task 6

In [25]:
# Calculate the mean price for each user.
ords_prods_merge['average_price'] = ords_prods_merge.groupby('user_id')['prices'].transform(np.mean)

In [26]:
# Assign a spending flag based on the average price.
ords_prods_merge.loc[ords_prods_merge['average_price'] < 10, 'spender_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['average_price'] >= 10, 'spender_flag'] = 'High spender'

In [27]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour  \
0   2539329        1             1                   2           8   
1   2398795        1             2                   3           7   
2    473747        1             3                   3          12   
3   2254736        1             4                   4           7   
4    431534        1             5                   4          15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  ...  _merge    price_range_loc     busiest day  \
0         Soda  ...    both  Mid-range product  Regularly busy   
1         Soda  ...    both  Mid-range product  Regularly busy   
2         Soda  ...    both  Mid-range product  Regularly busy   
3         Soda  ...    both  Mid-range product      Least busy   
4         Soda  ...    both  Mid-range product      Least busy   

      busiest_day busiest_slowest_days busiest_period_of_day max_order  \
0  Regularly busy       Regularly busy        Average orders        10   
1  Regularly busy         Slowest days        Average orders        10   
2  Regularly busy         Slowest days           Most orders        10   
3      Least busy         Slowest days        Average orders        10   
4      Least busy         Slowest days           Most orders        10   

   loyalty_flag average_price  spender_flag  
0  New customer      6.367797   Low spender  
1  New customer      6.367797   Low spender  
2  New customer      6.367797   Low spender  
3  New customer      6.367797   Low spender  
4  New customer      6.367797   Low spender  

[5 rows x 23 columns]

In [33]:
# Check the first few entries to verify the flags.
print(ords_prods_merge[['user_id', 'average_prices', 'spender_flag']].head())

   user_id    prices spender_flag
0        1  6.367797  Low spender
1        1  6.367797  Low spender
2        1  6.367797  Low spender
3        1  6.367797  Low spender
4        1  6.367797  Low spender


In [29]:
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low spender     31770614
High spender      634245
Name: count, dtype: int64

# step 7

In [30]:
# Calculate the median days since prior order for each user.
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby('user_id')['days_since_prior_order'].transform('median')


In [31]:
# Assign an order frequency flag based on the median days since prior order.
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] > 10) & (ords_prods_merge['median_days_since_prior_order'] <= 20), 'order_frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'order_frequency_flag'] = 'Frequent customer'


In [32]:
# Check the first few entries to verify the flags.
print(ords_prods_merge[['user_id', 'median_days_since_prior_order', 'order_frequency_flag']].head())

   user_id  median_days_since_prior_order order_frequency_flag
0        1                           20.0     Regular customer
1        1                           20.0     Regular customer
2        1                           20.0     Regular customer
3        1                           20.0     Regular customer
4        1                           20.0     Regular customer


In [34]:
ords_prods_merge['order_frequency_flag'].value_counts(dropna = False)

order_frequency_flag
Frequent customer        22790523
Regular customer          6927608
Non-frequent customer     2686728
Name: count, dtype: int64

# 8. export to csv

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_new.pkl'))